<a href="https://colab.research.google.com/github/cdo7456/BusinessIntelligenceProjects/blob/main/RSME%2C%20Linear%20Regression%20and%20Grid%20Search.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Data Cleaning, Description and Preparation

In [ ]:
# Importing the libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Importing 'Cancer' dataset
path = pd.read_csv('/content/drive/MyDrive/MGIS 355 (Spring 2023)/datasets/Cancer.csv')
cancer_df = path

In [ ]:
path.head()

countyCode State  PovertyEst  medIncome            Name  popEst2015  \
0        1001    AL        7204      54366  Autauga County       55347   
1        1003    AL       25696      49626  Baldwin County      203709   
2        1005    AL        5943      34971  Barbour County       26489   
3        1007    AL        3666      39546     Bibb County       22583   
4        1009    AL       10000      45567   Blount County       57673   

   incidenceRate  avgAnnCount recentTrend fiveYearTrend  deathRate  \
0          475.0        266.0      stable          -0.5      178.3   
1          454.6       1072.0      stable          -1.3      174.3   
2          477.5        155.0      stable          -1.4      192.7   
3          494.6        125.0      stable           3.3      212.4   
4          429.9        291.0      stable          -1.7      175.4   

   avgDeathsPerYear recTrend  
0                99  falling  
1               424  falling  
2                62  falling  
3                53   stable  
4               120  falling

In [ ]:
# Genereal information about the dataset, to show number of rows and columns
cancer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3072 entries, 0 to 3071
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   countyCode        3072 non-null   int64  
 1   State             3072 non-null   object 
 2   PovertyEst        3072 non-null   int64  
 3   medIncome         3072 non-null   int64  
 4   Name              3072 non-null   object 
 5   popEst2015        3072 non-null   int64  
 6   incidenceRate     3072 non-null   float64
 7   avgAnnCount       3072 non-null   float64
 8   recentTrend       3072 non-null   object 
 9   fiveYearTrend     3072 non-null   object 
 10  deathRate         3072 non-null   float64
 11  avgDeathsPerYear  3072 non-null   int64  
 12  recTrend          3072 non-null   object 
dtypes: float64(3), int64(5), object(5)
memory usage: 312.1+ KB


In [ ]:
cancer_df.describe()

countyCode    PovertyEst      medIncome    popEst2015  incidenceRate  \
count   3072.000000  3.072000e+03    3072.000000  3.072000e+03    3072.000000   
mean   30342.512695  1.567978e+04   47091.269206  1.045470e+05     448.287419   
std    15092.224711  5.614675e+04   12079.964480  3.325959e+05      54.481622   
min     1001.000000  9.100000e+01   22640.000000  8.270000e+02     201.300000   
25%    18180.500000  1.889500e+03   38876.250000  1.173900e+04     420.475000   
50%    29144.000000  4.435500e+03   45201.000000  2.693200e+04     453.550000   
75%    45053.500000  1.086550e+04   52509.250000  6.971500e+04     480.900000   
max    56045.000000  1.863025e+06  125635.000000  1.017029e+07    1206.900000   

        avgAnnCount    deathRate  avgDeathsPerYear  
count   3072.000000  3072.000000       3072.000000  
mean     612.505632   178.619954        187.769206  
std     1422.851775    27.732254        506.491882  
min        6.000000    59.700000          3.000000  
25%       76.000000   161.200000         28.000000  
50%      173.000000   178.150000         62.000000  
75%      522.250000   195.200000        150.000000  
max    38150.000000   362.800000      14010.000000

In [ ]:
# Shape of data
print(cancer_df.shape)

# Display column names
print(cancer_df.columns)

(3072, 13)
Index(['countyCode', 'State', 'PovertyEst', 'medIncome', 'Name', 'popEst2015',
       'incidenceRate', 'avgAnnCount', 'recentTrend', 'fiveYearTrend',
       'deathRate', 'avgDeathsPerYear', 'recTrend'],
      dtype='object')


In [ ]:
# Data Correlation
print(cancer_df.corr())

                  countyCode  PovertyEst  medIncome  popEst2015  \
countyCode          1.000000   -0.057949   0.060010   -0.056178   
PovertyEst         -0.057949    1.000000   0.116401    0.968736   
medIncome           0.060010    0.116401   1.000000    0.238723   
popEst2015         -0.056178    0.968736   0.238723    1.000000   
incidenceRate      -0.074558    0.013413  -0.002949    0.023981   
avgAnnCount        -0.094373    0.888021   0.269430    0.926633   
deathRate          -0.041023   -0.084420  -0.430815   -0.123186   
avgDeathsPerYear   -0.060316    0.943291   0.224364    0.976884   

                  incidenceRate  avgAnnCount  deathRate  avgDeathsPerYear  
countyCode            -0.074558    -0.094373  -0.041023         -0.060316  
PovertyEst             0.013413     0.888021  -0.084420          0.943291  
medIncome             -0.002949     0.269430  -0.430815          0.224364  
popEst2015             0.023981     0.926633  -0.123186          0.976884  
incidenceRate   

<ipython-input-41-eab8db69df66>:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  print(cancer_df.corr())


In [ ]:
# Identifying missing values
type(path)
cancer_df.isnull()

countyCode  State  PovertyEst  medIncome   Name  popEst2015  \
0          False  False       False      False  False       False   
1          False  False       False      False  False       False   
2          False  False       False      False  False       False   
3          False  False       False      False  False       False   
4          False  False       False      False  False       False   
...          ...    ...         ...        ...    ...         ...   
3067       False  False       False      False  False       False   
3068       False  False       False      False  False       False   
3069       False  False       False      False  False       False   
3070       False  False       False      False  False       False   
3071       False  False       False      False  False       False   

      incidenceRate  avgAnnCount  recentTrend  fiveYearTrend  deathRate  \
0             False        False        False          False      False   
1             False        False        False          False      False   
2             False        False        False          False      False   
3             False        False        False          False      False   
4             False        False        False          False      False   
...             ...          ...          ...            ...        ...   
3067          False        False        False          False      False   
3068          False        False        False          False      False   
3069          False        False        False          False      False   
3070          False        False        False          False      False   
3071          False        False        False          False      False   

      avgDeathsPerYear  recTrend  
0                False     False  
1                False     False  
2                False     False  
3                False     False  
4                False     False  
...                ...       ...  
3067             False     False  
3068             False     False  
3069             False     False  
3070             False     False  
3071             False     False  

[3072 rows x 13 columns]

In [ ]:
print(cancer_df.isnull().sum())

countyCode          0
State               0
PovertyEst          0
medIncome           0
Name                0
popEst2015          0
incidenceRate       0
avgAnnCount         0
recentTrend         0
fiveYearTrend       0
deathRate           0
avgDeathsPerYear    0
recTrend            0
dtype: int64


In [ ]:
cancer_df = cancer_df.dropna()

In [ ]:
print(cancer_df.isnull().sum())

countyCode          0
State               0
PovertyEst          0
medIncome           0
Name                0
popEst2015          0
incidenceRate       0
avgAnnCount         0
recentTrend         0
fiveYearTrend       0
deathRate           0
avgDeathsPerYear    0
recTrend            0
dtype: int64


There are 3072 rows, and 13 columns. There are no missing values in this dataset, so we do not need to drop or handle any missing values.

### Model Building for Incidence and Death Rate Prediction

In [ ]:
# Import libraries
import pandas as pd
from sklearn.model_selection import train_test_split, KFold, GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

In [ ]:
# check the number of unique counties
print(cancer_df['countyCode'].nunique())

# Since we're predicting by county we can drop the state column
# and county Name column since the countyCode can be used to
# uniquely identify a column

#print(cancer_df.head())
X = cancer_df.drop(['State', 'Name'], axis=1)


X = X.drop(['incidenceRate', 'deathRate'], axis=1)
Y = cancer_df[['incidenceRate', "deathRate"]]

3072


In [ ]:
# We check the unique values in recTrend and recentTrend

print(X['recentTrend'].unique())
print(X['recTrend'].unique())

['stable' 'falling' '*' 'rising' '�' '��']
['falling' 'stable' '**' 'rising']


In [ ]:
# recentTrend seems to contain some unknown values
# so we drop the column and take into account fiveYearTrend
# since fiveYearTrend is more specific and missing values
# can be substituted with 0.
# We use 0 to indicate there is no change in the trend
X = X.drop('recentTrend', axis=1)
X['fiveYearTrend'] = X['fiveYearTrend'].replace({'*': 0})
#print(X['fiveYearTrend'].unique())

In [ ]:
# We fill all unknown recentTrend values with 0.
# Making an assumption that whenever the value is missng
# the rate is stable

# We also fill missing values with 0 indicating stable


X['recTrend'] = X['recTrend'].map({'falling': -1, 'stable': 0, 'rising': 1, '**' : 0})
print(X.head())
print(Y.head())

   countyCode  PovertyEst  medIncome  popEst2015  avgAnnCount fiveYearTrend  \
0        1001        7204      54366       55347        266.0          -0.5   
1        1003       25696      49626      203709       1072.0          -1.3   
2        1005        5943      34971       26489        155.0          -1.4   
3        1007        3666      39546       22583        125.0           3.3   
4        1009       10000      45567       57673        291.0          -1.7   

   avgDeathsPerYear  recTrend  
0                99        -1  
1               424        -1  
2                62        -1  
3                53         0  
4               120        -1  
   incidenceRate  deathRate
0          475.0      178.3
1          454.6      174.3
2          477.5      192.7
3          494.6      212.4
4          429.9      175.4


In [ ]:
# split dataset into testing and training
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [ ]:
# # Import necessary libraries
# import pandas as pd
import numpy as np
# from sklearn.model_selection import KFold, GridSearchCV
# from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

#### Linear Regression

In [ ]:
cancer_df

countyCode State  PovertyEst  medIncome               Name  popEst2015  \
0           1001    AL        7204      54366     Autauga County       55347   
1           1003    AL       25696      49626     Baldwin County      203709   
2           1005    AL        5943      34971     Barbour County       26489   
3           1007    AL        3666      39546        Bibb County       22583   
4           1009    AL       10000      45567      Blount County       57673   
...          ...   ...         ...        ...                ...         ...   
3067       56037    WY        4346      72604  Sweetwater County       44626   
3068       56039    WY        1751      75348       Teton County       23125   
3069       56041    WY        2081      56800       Uinta County       20822   
3070       56043    WY        1030      50802    Washakie County        8328   
3071       56045    WY         901      55520      Weston County        7234   

      incidenceRate  avgAnnCount recentTrend fiveYearTrend  deathRate  \
0             475.0        266.0      stable          -0.5      178.3   
1             454.6       1072.0      stable          -1.3      174.3   
2             477.5        155.0      stable          -1.4      192.7   
3             494.6        125.0      stable           3.3      212.4   
4             429.9        291.0      stable          -1.7      175.4   
...             ...          ...         ...           ...        ...   
3067          403.5        155.0      stable           1.9      141.1   
3068          402.7         83.0      stable          -6.6      136.9   
3069          359.3         68.0      stable           4.1      132.8   
3070          448.0         50.0      stable          -3.8      182.3   
3071          397.1         37.0      stable           0.5      182.6   

      avgDeathsPerYear recTrend  
0                   99  falling  
1                  424  falling  
2                   62  falling  
3                   53   stable  
4                  120  falling  
...                ...      ...  
3067                49   stable  
3068                23  falling  
3069                24  falling  
3070                21   stable  
3071                17   stable  

[3072 rows x 13 columns]

In [ ]:
# Import necessary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold, GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
# Define the K-fold cross-validation object
cv = KFold(n_splits=5, shuffle=True, random_state=42)

In [ ]:
# Define the linear regression model
lr_model = LinearRegression()

In [ ]:
# Define the hyperparameter grid for tuning
param_grid = {
    'fit_intercept': [True, False],
    }

In [ ]:
# Define the GridSearchCV object for hyperparameter tuning
grid_search = GridSearchCV(lr_model, param_grid=param_grid, cv=cv, scoring='neg_mean_squared_error', n_jobs=-1)

In [ ]:
# Fit the GridSearchCV object to the training data
grid_search.fit(X_train, y_train)

GridSearchCV(cv=KFold(n_splits=5, random_state=42, shuffle=True),
             estimator=LinearRegression(), n_jobs=-1,
             param_grid={'fit_intercept': [True, False]},
             scoring='neg_mean_squared_error')

In [ ]:
# Print the best hyperparameters
print('Best hyperparameters:', grid_search.best_params_)

Best hyperparameters: {'fit_intercept': True}


In [ ]:
# Predict on the testing set
y_pred = grid_search.predict(X_test)

#####RSME Calculation for Linear Regression




In [ ]:
# Calculate the mean squared error and R2 score
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [ ]:
# Print the results
print('MSE:', mse)
print('R2:', r2)

MSE: 2043.0119810302242
R2: 0.14819971517126374


#### K-Nearest Neighbours

In [ ]:
# Define the features and target variables
#features = cancer_df.drop(['deathRate', 'countyCode', 'incidenceRate'], axis=1)
#target_death = cancer_df['deathRate']
#target_inc = cancer_df['incidenceRate']

In [ ]:
# Define the parameter grid to search over
param_grid = {
    'n_neighbors': [3, 5, 7, 10],
    'weights': ['uniform', 'distance'],
    'p': [1, 2],
    'algorithm': ['ball_tree', 'kd_tree']}

In [ ]:
# Initialize the K-Nearest Neighbours models
knn = KNeighborsRegressor(n_neighbors=5)
#knn_inc = KNeighborsRegressor(n_neighbors=5)

In [ ]:
# Initialize the RandomizedSearchCV objects
from sklearn.model_selection import train_test_split, RandomizedSearchCV
knn_cv = RandomizedSearchCV(knn, param_grid, cv=5, n_iter=10, random_state=42, n_jobs=-1)
#knn_cv = RandomizedSearchCV(knn_inc, param_grid, cv=5, n_iter=10, random_state=42, n_jobs=-1)

In [ ]:
# Fit the models to the training data using K-fold cross-validation
knn_cv.fit(X_train, y_train)
#knn_inc_cv.fit(X_train, y_inc_train)

RandomizedSearchCV(cv=5, estimator=KNeighborsRegressor(), n_jobs=-1,
                   param_distributions={'algorithm': ['ball_tree', 'kd_tree'],
                                        'n_neighbors': [3, 5, 7, 10],
                                        'p': [1, 2],
                                        'weights': ['uniform', 'distance']},
                   random_state=42)

In [ ]:
# Print the best hyperparameters for each model
print('Best hyperparameters for death rate model:', knn_cv.best_params_)
#print('Best hyperparameters for incidence rate model:', knn_inc_cv.best_params_)

Best hyperparameters for death rate model: {'weights': 'distance', 'p': 1, 'n_neighbors': 10, 'algorithm': 'kd_tree'}


In [ ]:
# Make predictions on the testing set using the best models
y_pred = knn_cv.best_estimator_.predict(X_test)
#y_inc_pred = knn_inc_cv.best_estimator_.predict(X_test)

#####RSME Calculation for KNN

In [ ]:
# Calculate the root mean squared error (RMSE) of the best models
rmse = mean_squared_error(y_test, y_pred, squared=False)
#rmse_inc = mean_squared_error(y_inc_test, y_inc_pred, squared=False)

In [ ]:
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [ ]:
# Print the RMSE scores
print('KNN RMSE:', rmse)
print('KNN R2:', r2)
#print('KNN Incidence Rate RMSE:', rmse_inc)

KNN RMSE: 41.43474579584809
KNN R2: 0.15281438082076731


#### Random Forest

In [ ]:
# Define the features and target variables
features = cancer_df.drop(['deathRate', 'countyCode', 'incidenceRate'], axis=1)
target_death = cancer_df['deathRate']
target_inc = cancer_df['incidenceRate']

In [ ]:
# Spilt the dataset into training and test sets
#X_train, X_test, y_death_train, y_death_test, y_inc_train, y_inc_test = train_test_split(features, target_death, target_inc, test_size=0.2, random_state=42)

In [ ]:
# Define the parameter grid to search over
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 10, None],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2],
    'max_features': ['sqrt']}

In [ ]:
# Initialize the Random Forest models
rf = RandomForestRegressor(n_estimators=100, random_state=42)
#rf_inc = RandomForestRegressor(n_estimators=100, random_state=42)

In [ ]:
# Initialize the RandomizedSearchCV objects
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.ensemble import RandomForestRegressor

rf_cv = RandomizedSearchCV(rf, param_grid, cv=5, n_iter=10, random_state=42)
#rf_inc_cv = RandomizedSearchCV(rf_inc, param_grid, cv=5, n_iter=10, random_state=42)

In [ ]:
# Fit the models to the training data using K-fold cross-validation
rf_cv.fit(X_train, y_train)
#rf_inc_cv.fit(X_train, y_inc_train)

RandomizedSearchCV(cv=5, estimator=RandomForestRegressor(random_state=42),
                   param_distributions={'max_depth': [3, 5, 10, None],
                                        'max_features': ['sqrt'],
                                        'min_samples_leaf': [1, 2],
                                        'min_samples_split': [2, 5],
                                        'n_estimators': [50, 100, 200]},
                   random_state=42)

In [ ]:
# Print the best hyperparameters for each model
print('Best hyperparameters for death rate model:', rf_cv.best_params_)
#print('Best hyperparameters for incidence rate model:', rf_inc_cv.best_params_)

Best hyperparameters for death rate model: {'n_estimators': 100, 'min_samples_split': 2, 'min_samples_leaf': 1, 'max_features': 'sqrt', 'max_depth': None}


In [ ]:
# Make predictions on the testing set using the best models
y_death_pred = rf_cv.best_estimator_.predict(X_test)
#y_inc_pred = rf_inc_cv.best_estimator_.predict(X_test)

#####RSME Calculation for Random Forest

In [ ]:
# Calculate the root mean squared error (RMSE) of the best models
rmse = mean_squared_error(y_test, y_death_pred, squared=False)
#rmse_inc = mean_squared_error(y_inc_test, y_inc_pred, squared=False)

In [ ]:
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [ ]:
# Print the RMSE scores of the model
print('Random Forest RMSE:', rmse)
print('Random Forest R2:', r2)
#print('Random Forest Incidence Rate RMSE:', rmse_inc)

Random Forest RMSE: 35.98548369951093
Random Forest R2: 0.15281438082076731


###Comparison of Prediction Results for Each Model

####Analysis

To compare the prediction results of each model, we can use metrics such as mean squared error (MSE), root mean squared error (RMSE), and R2 score. The lower the MSE and RMSE and the higher the R2 score, the better the model's performance.

Here are the performance metrics for each model:



>




> **Linear Regression Model:**
* RMSE: 2043.0119810302242
* R2: 0.14819971517126374


> **K-Nearest Neighbours model:**
* RMSE: 41.43474579584809
* R2: 0.15281438082076731


> **Random Forest model:**
* RMSE: 35.98548369951093
* R2: 0.15281438082076731


####Reccomendation

Based on these results, the **Random Forest model performed the best**, with the lowest MSE and RMSE for both incidence rate and death rate and the highest R2 score for death rate. However, it's worth noting that the KNN and Random Forest models had relatively similar performance, with the KNN model only slightly worse than the Random Forest model. The Linear Regression model performed the worst.

Professionals should use the Random Forest model because it provides accurate predictions for both incidence rate and death rate. However, it's always a good idea to consider multiple models and compare their performance before making any decisions.